### PR TITLE
Add support for the YD68/YD68v2 bluetooth module

### PR DESCRIPTION
**Pull request by izzy84075**
Probably not well integrated, but it does seem to work. Not sure if there's a way to hook this into the "normal" bluetooth stuff.

COMMAND+U switches between USB and Bluetooth modes. Bluetooth module can be disabled entirely with the YD68_XM04_ENABLE option in rules.mk.

Currently if the option is enabled, the bluetooth module is always left on, but there is the possibility to turn it on and off when toggling between USB and Bluetooth modes. Not sure if that's a better "default" thing.

### CODE REVIEW

**Comment by izzy84075** [comment]
I'm not sure what to do about that Travis error. Builds fine on my machine, and I'm not sure why the Travis build would differ for that one.

File: keyboards/yd68/xm04/main.c [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    clock_prescale_set(clock_div_1);
+
+    // Leonardo needs. Without this USB device is not recognized.
+    USB_Disable();
+
+    USB_Init();
+
+    // for Console_Task
+    USB_Device_EnableSOFEvents();
+    print_set_sendchar(sendchar_func);
```
Comment (drashna) [inline keyboards/yd68/xm04/main.c]: I think this should be `print_set_sendchar(sendchar_func( ));`

**Comment by mechmerlin** [comment]
@izzy84075 you're my hero! I've been waiting for someone to do this!

File: keyboards/yd68/xm04/main.c (hunk 1 above)
Comment (izzy84075) [reply keyboards/yd68/xm04/main.c]: It does take a function pointer, so print_set_sendchar(sendchar_func); is correct. keyboards/hhkb/rn42/main.c uses the same form.

**Comment by izzy84075** [comment]
> 
> 
> @izzy84075 you're my hero! I've been waiting for someone to do this!

I've gotten too used to QMK on my Tada68 at work(I also added support for Velocikey ( #3754 ) on LED backlighting specifically for that keyboard), I couldn't break away when I got a keyboard for home, but I wanted wireless capability. So I made it happen. :) Glad somebody else is happy about it already. :)

File: keyboards/yd68/xm04/main.c [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#include "action.h"
+#include "action_util.h"
+#include "wait.h"
+#include "suart.h"
+#include "suspend.h"
+extern bool dozing;
+extern bool sleeping;
+extern bool force_usb;
+
+static int8_t sendchar_func(uint8_t c)
```
Comment (mechmerlin) [inline keyboards/yd68/xm04/main.c]: This is a statically defined routine that is only available in this `.c` file, but is not used. You can remove this.

Comment (izzy84075) [reply keyboards/yd68/xm04/main.c]: Will removing this and the reference to it below still allow debug output on the hid_listen console, if needed? It's something related to that, I know, but I'm not quite sure how all of it's tied together.

**Comment by izzy84075** [comment]
So I've been doing some more testing, now that I've actually got a battery in my YD68, and there's some more work that needs done on this, I think. When Bluetooth is enabled with the option in rules.mk, the RGB LEDs don't work, and after that last commit, neither does the debug console. So I need to do some more looking at the "real' main(), and implement the xm04 main closer to that, I think.

**Comment by mechmerlin** [comment]
@izzy84075 has there been any more progress on this?

**Comment by drashna** [comment]
Any updates on this?

**Comment by izzy84075** [comment]
Ack, sorry... Completely forgot... IIRC, I got to the point where I was having to look into how to add the bluetooth driver to the "core" of qmk, because the way I was doing it basically removed everything beyond basic functionality...

**Comment by drashna** [comment]
I think you'd need to add stuff to `tmk_core` specifically.  Specifically, `tmk_core/protocol` I think.

**Comment by skullydazed** [comment]
@izzy84075 Are you still working on this feature?

**Comment by izzy84075** [comment]
Not very often, no. I haven't had enough of a chance to dig into the protocol stuff in QMK to figure out how to hook in a new interface.

**Comment by drashna** [comment]
@izzy84075 well, there is the bluefruit stuff, that may be a very good place to start. 

In fact, I think that a lot of this would be just moving the stuff into a folder, with some additional configuration.

**Comment by skullydazed** [comment]
In an effort to clean up our queue I'm going to close this PR. However, we'd still like this feature to be included once it's ready! Please feel free to open a new PR when you are ready.

**Pull request closed without merging**